src/main/java/org/icatproject/core/entity/Datafile.java

```diff
@@ -195,25 +202,111 @@ public void setSourceDatafiles(List<RelatedDatafile> sourceDatafiles) {
 
 	@Override
 	public void getDoc(JsonGenerator gen) {
-		StringBuilder sb = new StringBuilder(name);
+		SearchApi.encodeString(gen, "name", name);
 		if (description != null) {
-			sb.append(" " + description);
+			SearchApi.encodeString(gen, "description", description);
+		}
+		if (location != null) {
+			SearchApi.encodeString(gen, "location", location);
 		}
 		if (doi != null) {
-			sb.append(" " + doi);
+			SearchApi.encodeString(gen, "doi", doi);
 		}
+		if (fileSize != null) {
+			SearchApi.encodeLong(gen, "fileSize", fileSize);
+		} else {
+			SearchApi.encodeLong(gen, "fileSize", 0L);
+		}
+		SearchApi.encodeLong(gen, "fileCount", 1L); // Always 1, but makes sorting on fields consistent
 		if (datafileFormat != null) {
-			sb.append(" " + datafileFormat.getName());
+			datafileFormat.getDoc(gen);
 		}
-		LuceneApi.encodeTextfield(gen, "text", sb.toString());
 		if (datafileModTime != null) {
-			LuceneApi.encodeStringField(gen, "date", datafileModTime);
+			SearchApi.encodeLong(gen, "date", datafileModTime);
 		} else if (datafileCreateTime != null) {
-			LuceneApi.encodeStringField(gen, "date", datafileCreateTime);
+			SearchApi.encodeLong(gen, "date", datafileCreateTime);
 		} else {
-			LuceneApi.encodeStringField(gen, "date", modTime);
+			SearchApi.encodeLong(gen, "date", modTime);
+		}
+		SearchApi.encodeString(gen, "id", id);
+		if (dataset != null) {
+			SearchApi.encodeString(gen, "dataset.id", dataset.id);
+			SearchApi.encodeString(gen, "dataset.name", dataset.getName());
+			Sample sample = dataset.getSample();
+			if (sample != null) {
+				sample.getDoc(gen);
+			}
+			Investigation investigation = dataset.getInvestigation();
+			if (investigation != null) {
+				SearchApi.encodeString(gen, "investigation.id", investigation.id);
+				SearchApi.encodeString(gen, "investigation.name", investigation.getName());
+				SearchApi.encodeString(gen, "visitId", investigation.getVisitId());
```

Comment:
We have to prefix with `investigation.` for `name`, `id` etc. because these are common to multiple entities (i.e. we need to distinguish between an `investigation.name` and the Datafile's own `name`.
We don't have to do this for `visitId`, because it's the only field with that name (but we could). The benefit of leaving it as `visitId` is that if you perform a keyword search like `visitId:AB1234`, this will match across all three entity types because they all have the Investigation's field encoded with the name `visitId`. Being able to search Datafiles using the `visitId` was something that Diamond specifically wanted, so this ensures you get that without having to do something less intuitive like `investigation.visitId:AB1234`.

Reply:
That makes sense, thank you for your explanation.